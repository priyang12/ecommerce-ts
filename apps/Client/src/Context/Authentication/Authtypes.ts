export const SetAuth = "SetAuth";
export const LOGIN_SUCCESS = "LOGIN_SUCCESS";
export const MailSEND_SUCCESS = "MailSEND_SUCCESS";
export const RESET_PASSWORD_SUCCESS = "RESET_PASSWORD_SUCCESS";
export const UPDATE_USER = "UPDATE_USER";
export const LOAD_USER = "LOAD_USER";
export const SET_LOADING = "SET_LOADING";
export const AUTH_ERROR = "AUTH_ERROR";
export const LOG_OUT = "LOG_OUT";
