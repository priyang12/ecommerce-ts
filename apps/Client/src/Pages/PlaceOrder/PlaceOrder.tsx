import { Navigate as Redirect, useNavigate } from "react-router-dom";
import { Helmet } from "react-helmet-async";
import {
  StyledHeader as Header,
  StyledOrderSummary as OrderSummary,
  StyledOrderSummaryBody as OrderSummaryBody,
  StyledOrderSummaryItem as OrderSummaryItem,
  StyledParagraph as Paragraph,
  StyledItemContainer,
  StyledPlaceOrder,
} from "./StyledPlaceOrder";
import { useLoadCartQuery } from "../../API/CartAPI";
import Spinner from "../../Components/Spinner";
import { useCheckout } from "../../Context/CheckoutContext/CheckoutContext";
import { StyledCheckoutContainer } from "../../Components/UI/CheckoutContainer";
import { SubmitButton } from "../../Components/UI/FormControl";
import OrderItem from "./ProductList";

import { css } from "@linaria/core";

const containerWidth = css`
  max-width: 80vw;
`;

/**
 * PlaceOrder Component
 *
 * Displays the final step of the checkout process where users can review
 * shipping details, payment method, and a summary of their order before placing it.
 *
 *
 * ## Route
 * - `/checkout/placeOrder`
 * @component
 *
 * @returns {JSX.Element} A React component that renders the place order page.
 *
 * @requires
 * - A valid address in checkout state or localStorage
 * - A selected payment method
 * - A non-empty cart (from `useLoadCartQuery`)
 *
 * @behavior
 * - Redirects to `/checkout/address` if no address
 * - Redirects to `/checkout/paymentMethod` if no payment method
 * - Redirects to `/` if the cart is empty
 *
 * @todo
 * - Replace local calculation logic with backend-confirmed values if applicable.
 * - Add validation/error states for failed order submissions
 * - Animate transitions or confirmation modals
 */
const PlaceOrder = () => {
  const Navigate = useNavigate();
  const {
    state: { address, payMethod },
    dispatch,
  } = useCheckout();

  const { data: cart, isLoading } = useLoadCartQuery();

  const ProductsAmount = cart
    ? cart.products.reduce((pre, current) => {
        return (pre += current.product.price * current.qty);
      }, 0)
    : 0;
  const roundedProductAmount = ProductsAmount.toFixed(2);
  const ShippingAmount = ProductsAmount > 500 ? 0 : 100;
  const TaxAmount = +(0.15 * ProductsAmount).toFixed(2);

  const TotalAmount = +(ProductsAmount + ShippingAmount + TaxAmount).toFixed(2);

  const PlaceTheOrder = (e: React.ChangeEvent<HTMLFormElement>) => {
    e.preventDefault();
    const Order = {
      orderItems: cart?.products,
      shippingAddress: address,
      paymentMethod: payMethod,
      itemsPrice: ProductsAmount,
      taxPrice: TaxAmount,
      shippingPrice: ShippingAmount,
      totalPrice: Math.round(TotalAmount),
    };

    dispatch({
      type: "SET_ORDER",
      payload: Order,
    });

    Navigate("/checkout/paypal");
  };

  if (!address) return <Redirect to="/checkout/address" />;
  if (!payMethod) return <Redirect to="/checkout/paymentMethod" />;

  if (isLoading) return <Spinner />;

  if (cart && cart.products?.length === 0) return <Redirect to="/" />;

  return (
    <StyledCheckoutContainer className={containerWidth}>
      <Helmet>
        <title>Place Order</title>
        <meta name="description" content="Place Order" />
      </Helmet>
      <StyledPlaceOrder>
        <div>
          <div>
            <Header>SHIPPING </Header>
            <Paragraph>
              Address: {address.address} , {address.city} ,{address.postalcode},
            </Paragraph>
            <Header>PAYMENT METHOD</Header>
            <Paragraph>Method: {payMethod}</Paragraph>
          </div>

          <div className="order-details">
            <Header>ORDER ITEMS</Header>
            <StyledItemContainer>
              {cart?.products.map((item) => (
                <OrderItem Cart={item} key={item._id} styledWidth="80%" />
              ))}
            </StyledItemContainer>
          </div>
        </div>
        <OrderSummary>
          <form onSubmit={PlaceTheOrder} data-testid="PlaceOrder">
            <OrderSummaryBody>
              <Header>ORDER SUMMARY</Header>
              <OrderSummaryItem>
                Items Cost : <span> {roundedProductAmount}</span>
              </OrderSummaryItem>
              <OrderSummaryItem>
                Shipping Cost :
                <span data-testid="ShippingCost">{ShippingAmount}</span>
              </OrderSummaryItem>
              <OrderSummaryItem>
                Tax Cost :<span data-testid="TaxCost">{TaxAmount}</span>
              </OrderSummaryItem>
              <OrderSummaryItem>
                Total Cost :
                <span data-testid="TotalAmount">{Math.round(TotalAmount)}</span>
              </OrderSummaryItem>
            </OrderSummaryBody>
            <SubmitButton type="submit" value="PlaceOrder" />
          </form>
        </OrderSummary>
      </StyledPlaceOrder>
    </StyledCheckoutContainer>
  );
};

export default PlaceOrder;
