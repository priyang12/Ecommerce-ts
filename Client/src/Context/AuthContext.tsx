import { AuthState, IContextModel } from './interfaces';
import { createContext, useReducer } from 'react';
import { AuthReducer } from './AuthReducer';

let Token = null;
if (localStorage.token) Token = localStorage.token;

const initialState: AuthState = {
  loading: false,
  err: null,
  token: Token,
  user: null,
};

export const AuthContext = createContext({} as IContextModel);

export const AuthProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(AuthReducer, initialState);
  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};
