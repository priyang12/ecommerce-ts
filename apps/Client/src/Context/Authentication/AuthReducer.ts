import { User } from "../../Types/interfaces";
import { AuthState } from "./AuthContext";
import {
  LOGIN_SUCCESS,
  AUTH_ERROR,
  LOAD_USER,
  SET_LOADING,
  LOG_OUT,
  UPDATE_USER,
  MailSEND_SUCCESS,
  RESET_PASSWORD_SUCCESS,
} from "./Authtypes";

type AuthPayload = {
  [LOGIN_SUCCESS]: string | null;
  [AUTH_ERROR]: string | null;
  [LOAD_USER]: User | null;
  [UPDATE_USER]: User;
  [SET_LOADING]: boolean;
  [LOG_OUT]: null;
  [MailSEND_SUCCESS]: string;
  [RESET_PASSWORD_SUCCESS]: string;
};

type ActionMap<M extends { [index: string]: any }> = {
  [Key in keyof M]: M[Key] extends undefined
    ? {
        type: Key;
      }
    : {
        type: Key;
        payload: M[Key];
      };
};

export type AuthActions = ActionMap<AuthPayload>[keyof ActionMap<AuthPayload>];

export const AuthReducer = (state: AuthState, action: AuthActions) => {
  switch (action.type) {
    case LOGIN_SUCCESS:
      return {
        ...state,
        token: action.payload,
        loading: false,
      };
    case LOAD_USER:
    case UPDATE_USER:
      return {
        ...state,
        user: action.payload,
        loading: false,
      };
    case MailSEND_SUCCESS:
    case RESET_PASSWORD_SUCCESS:
      return {
        ...state,
        loading: false,
        alert: {
          show: true,
          message: action.payload,
          type: "success",
        },
      };
    case SET_LOADING:
      return {
        ...state,
        loading: action.payload,
      };
    case AUTH_ERROR:
      return {
        ...state,
        loading: false,
        err: action.payload,
      };

    case LOG_OUT:
      return {
        ...state,
        token: null,
        loading: false,
        user: null,
        err: null,
      };
  }
};
