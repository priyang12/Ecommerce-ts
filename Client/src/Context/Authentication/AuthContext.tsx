import { AuthState, IContextModel } from "./interfaces";
import { createContext, useReducer } from "react";
import { AuthReducer } from "./AuthReducer";

let Token = null;
if (localStorage.token) Token = localStorage.token;

const user = sessionStorage.getItem("User");

const initialState: AuthState = {
  loading: true,
  err: null,
  token: Token,
  user: user && JSON.parse(user),
};

export const AuthContext = createContext({} as IContextModel);

export const AuthProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(AuthReducer, initialState);
  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};
