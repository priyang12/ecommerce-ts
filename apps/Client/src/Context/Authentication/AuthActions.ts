import {
  LOGIN_SUCCESS,
  AUTH_ERROR,
  LOAD_USER,
  SET_LOADING,
  LOG_OUT,
  UPDATE_USER,
  MailSEND_SUCCESS,
  RESET_PASSWORD_SUCCESS,
} from "./Authtypes";

import axios, { AxiosError } from "axios";
import { AuthActions } from "./AuthReducer";
import React from "react";

export const loadUser = async (
  token: any,
  dispatch: React.Dispatch<AuthActions>
) => {
  if (token) {
    try {
      setLoading(dispatch);
      const { data }: any = await axios.get("/api/users", {
        headers: {
          "x-auth-token": token,
        },
      });
      sessionStorage.setItem("User", JSON.stringify(data));
      dispatch({
        type: LOAD_USER,
        payload: data,
      });
    } catch (err: any | AxiosError) {
      let ErrorMessage = "Server Error Try Again Later";
      if (err as AxiosError) {
        ErrorMessage = err.response.data?.msg;
      }
      Logout(dispatch);
      dispatch({
        type: AUTH_ERROR,
        payload: ErrorMessage,
      });
    }
  }
};
export const LoginUser = async (
  user: any,
  dispatch: React.Dispatch<AuthActions>
) => {
  try {
    setLoading(dispatch);
    const { data }: any = await axios.post("/api/users/login", user);
    dispatch({
      type: LOGIN_SUCCESS,
      payload: data.token,
    });
    localStorage.setItem("token", data.token);

    // loadUser(data.token, dispatch);
  } catch (err: any | AxiosError) {
    let ErrorMessage = "Server Error Try Again Later";
    if (err as AxiosError) {
      ErrorMessage = err.response?.data?.msg;
    }

    dispatch({
      type: AUTH_ERROR,
      payload: ErrorMessage,
    });
  }
};
export const RegisterUser = async (
  user: any,
  dispatch: React.Dispatch<AuthActions>
) => {
  try {
    setLoading(dispatch);
    const { data }: any = await axios.post("/api/users/register", user);
    dispatch({
      type: LOGIN_SUCCESS,
      payload: data.token,
    });
    localStorage.setItem("token", data.token);

    loadUser(data.token, dispatch);
  } catch (err: any | AxiosError) {
    let ErrorMessage = "Server Error Try Again Later";
    if (err as AxiosError) {
      ErrorMessage = err.response.data?.msg;
    }
    dispatch({
      type: AUTH_ERROR,
      payload: ErrorMessage,
    });
  }
};

export const UpdateUser = async (
  dispatch: React.Dispatch<AuthActions>,
  user: any
) => {
  try {
    setLoading(dispatch);
    const { data }: any = await axios.put("/api/users/update", user);
    dispatch({
      type: UPDATE_USER,
      payload: data,
    });
  } catch (err: any | AxiosError) {
    let ErrorMessage = "Server Error Try Again Later";
    if (err as AxiosError) {
      ErrorMessage = err.response.data.message;
    }
    dispatch({
      type: AUTH_ERROR,
      payload: ErrorMessage,
    });
  }
};

export const UpdatePassword = async (
  dispatch: React.Dispatch<AuthActions>,
  password: any,
  password2: any
) => {
  try {
    setLoading(dispatch);
    const { data }: any = await axios.patch("/api/users/resetpassword", {
      password,
      password2,
    });
    dispatch({
      type: RESET_PASSWORD_SUCCESS,
      payload: data.message,
    });
  } catch (err: any | AxiosError) {
    let ErrorMessage = "Server Error Try Again Later";

    if (err as AxiosError) {
      ErrorMessage = err.response?.data?.msg;
    }
    dispatch({
      type: AUTH_ERROR,
      payload: ErrorMessage,
    });
  }
};

export const RecoverPassword = async (
  dispatch: React.Dispatch<AuthActions>,
  email: string
) => {
  try {
    setLoading(dispatch);
    const { data }: any = await axios.post("/api/users/recoverMail", { email });
    dispatch({
      type: MailSEND_SUCCESS,
      payload: data.message,
    });
  } catch (err: any | AxiosError) {
    let ErrorMessage = "Server Error Try Again Later";

    if (err as AxiosError) {
      ErrorMessage = err.response.data?.msg;
    }
    dispatch({
      type: AUTH_ERROR,
      payload: ErrorMessage,
    });
  }
};

export const Logout = (dispatch: React.Dispatch<AuthActions>) => {
  try {
    localStorage.clear();
    sessionStorage.clear();
    dispatch({
      type: LOG_OUT,
      payload: null,
    });
  } catch {
    dispatch({
      type: AUTH_ERROR,
      payload: "Server Error Reload",
    });
  }
};
export const setLoading = (dispatch: React.Dispatch<AuthActions>) => {
  dispatch({
    type: SET_LOADING,
    payload: true,
  });
};

export const StopLoading = (dispatch: React.Dispatch<AuthActions>) => {
  dispatch({
    type: SET_LOADING,
    payload: false,
  });
};
