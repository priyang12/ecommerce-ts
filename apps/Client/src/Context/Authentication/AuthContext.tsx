import React, { createContext, useReducer } from "react";
import { AuthActions, AuthReducer } from "./AuthReducer";
import { User } from "../../Types/interfaces";

export interface IContextModel {
  state: AuthState;
  dispatch: React.Dispatch<AuthActions>;
}
interface alert {
  show: boolean;
  message: string;
  type: string;
}

export type AuthUser = Omit<User, "password" | "updatedAt" | "createdAt">;

export interface AuthState {
  loading: boolean;
  err: string | null;
  token: string | null;
  user: AuthUser | null;
  alert: alert | null;
}

let Token = null;
if (localStorage.token) Token = localStorage.token;

const user = sessionStorage.getItem("User");

const initialState: AuthState = {
  loading: false,
  err: null,
  token: Token,
  user: user && JSON.parse(user),
  alert: null,
};

export const AuthContext = createContext({} as IContextModel);
AuthContext.displayName = "AuthContext";
export const useAuth = () => {
  return React.useContext(AuthContext);
};

export const AuthProvider = ({ children }: { children: any }) => {
  const [state, dispatch] = useReducer(AuthReducer, initialState);
  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};
