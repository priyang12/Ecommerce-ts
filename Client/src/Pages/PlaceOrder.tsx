import React, { useContext, useEffect } from "react";
import { Redirect, useHistory } from "react-router";
import TimeoutBtn from "../Components/TimeoutBtn";
import ProductList from "../Components/ProductList";
import Navigators from "../Components/Navigators";
import { Address, Cart } from "../interfaces";
import { AuthContext } from "../Context/Authentication/AuthContext";
import { StyledPaymentContainer } from "../Components/StyledComponents/StyledPayment";

import {
  StyledHeader,
  StyledOrderSummary,
  StyledOrderSummaryBody,
  StyledOrderSummaryItem,
  StyledPlaceOrder,
} from "./StyledPages/StyledPlaceOrder";

const PlaceOrder = () => {
  const history = useHistory();
  const { state } = useContext(AuthContext);
  const { token } = state;

  const Cart: Cart[] = localStorage.Cart && JSON.parse(localStorage.Cart);
  const Address: Address =
    localStorage.address && JSON.parse(localStorage.address);
  const PayMethod = localStorage.payMethod;
  const ProductsAmount =
    localStorage.ProductsAmount && JSON.parse(localStorage.ProductsAmount);

  let ExtraAmount: number = 0;

  useEffect(() => {
    if (!token) {
      history.push("/");
    }
  }, [token, history]);

  const addDecimals = (num: number) => {
    const result = Number((Math.round(num * 100) / 100).toFixed(2));
    ExtraAmount += result;
    return result;
  };
  const PlaceTheOrder = (e: React.ChangeEvent<HTMLFormElement>) => {
    e.preventDefault();
    const shipping = addDecimals(ProductsAmount > 500 ? 0 : 100);
    const Tax = addDecimals(ProductsAmount * 0.1);
    const Order = {
      orderItems: Cart,
      shippingAddress: Address,
      paymentMethod: PayMethod,
      itemsPrice: ProductsAmount,
      taxPrice: 0.1 * ProductsAmount,
      shippingPrice: ProductsAmount > 500 ? 0 : 100,
      totalPrice: Math.round(shipping + Tax + ProductsAmount),
    };
    localStorage.setItem("order", JSON.stringify(Order));
    history.push("/PayPal");
  };

  if (!Cart || !Address || !PayMethod || !ProductsAmount)
    return <Redirect to="/cart" />;

  if (Cart?.length === 0) return <Redirect to="/" />;

  return (
    <StyledPaymentContainer theme={{ maxWidth: "80vw" }}>
      <Navigators />
      <StyledPlaceOrder>
        <div className="OrderDetails">
          <div className="detail">
            <StyledHeader>SHIPPING </StyledHeader>
            <p>
              Address: {Address.homeAddress} , {Address.city} ,
              {Address.postalCode},
            </p>
          </div>
          <div className="detail">
            <StyledHeader>PAYMENT METHOD</StyledHeader>
            <p>Method: {PayMethod}</p>
          </div>
          <div className="order-details">
            <StyledHeader>ORDER ITEMS</StyledHeader>
            <ul className="order-list">
              {Cart.map((item) => (
                <ProductList Cart={item} key={item._id} styledWidth="80%" />
              ))}
            </ul>
          </div>
        </div>
        <StyledOrderSummary>
          <form onSubmit={PlaceTheOrder} data-testid="PlaceOrder">
            <StyledOrderSummaryBody>
              <StyledHeader>ORDER SUMMARY</StyledHeader>
              <StyledOrderSummaryItem>
                Items Cost : <span> {ProductsAmount}</span>
              </StyledOrderSummaryItem>
              <StyledOrderSummaryItem>
                Shipping Cost :
                <span data-testid="ShippingCost">
                  {addDecimals(ProductsAmount > 500 ? 0 : 100)}
                </span>
              </StyledOrderSummaryItem>
              <StyledOrderSummaryItem>
                Tax Cost :
                <span data-testid="TaxCost">
                  {addDecimals(0.1 * ProductsAmount)}
                </span>
              </StyledOrderSummaryItem>
              <StyledOrderSummaryItem>
                Total Cost :
                <span data-testid="TotalAmount">
                  {Math.round(ExtraAmount + ProductsAmount)}
                </span>
              </StyledOrderSummaryItem>
            </StyledOrderSummaryBody>

            <TimeoutBtn Time={4000} className="btn" FormValue="PlaceOrder" />
          </form>
        </StyledOrderSummary>
      </StyledPlaceOrder>
    </StyledPaymentContainer>
  );
};

export default PlaceOrder;
