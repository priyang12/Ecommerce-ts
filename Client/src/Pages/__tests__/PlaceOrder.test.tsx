import { render, screen, act } from "@testing-library/react";
import { Router } from "react-router-dom";
import { AuthContext } from "../../Context/Authentication/AuthContext";
import { createMemoryHistory, MemoryHistory } from "history";
import PlaceOrder from "../PlaceOrder";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

let History: MemoryHistory<unknown>;

// Auth Context

const state = {
  loading: false,
  err: null,
  token: null,
  user: {
    name: "Priyang",
    isAdmin: false,
    email: "asdas",
    id: "asdasd",
    createdAt: "asasd",
  },
};
const dispatch = jest.fn();

//  Local Storage Value
///////////////////////////////xxxxxxxxxxxxxxxxxxxxxxx//////////////////////////////
///////////////////////////////xxxxxxxxxxxxxxxxxxxxxxx//////////////////////////////
const Cart = [
  {
    _id: "617028b4ae931d0004f7d964",
    product: {
      price: 89.99,
      countInStock: 2,
      _id: "60d5e622e5179e2bb44bd838",
      name: "Airpods Wireless Bluetooth Headphones",
      image: "/Photos/image-1627384388351.webp",
    },
    qty: 2,
  },
  {
    _id: "61705624b54854000494b5ce",
    product: {
      price: 929.99,
      countInStock: 5,
      _id: "60d5e622e5179e2bb44bd83c",
      name: "Logtech mouse",
      image: "/Photos/image-1627385386692.webp",
    },
    qty: 1,
  },
];

const address = {
  homeAddress: "202,Pipload",
  city: "Surat",
  postalCode: "456123",
  country: "India",
};

const Method = "PayPal or Credit Card";

const productsAmount = 2000;

///////////////////////////////xxxxxxxxxxxxxxxxxxxxxxx//////////////////////////////
///////////////////////////////xxxxxxxxxxxxxxxxxxxxxxx//////////////////////////////

let ExtraAmount = 0;
const addDecimals = (num: number) => {
  const result = Number((Math.round(num * 100) / 100).toFixed(2));
  ExtraAmount += result;
  return result;
};

beforeAll(() => {
  History = createMemoryHistory();
});

const Setup = () => {
  localStorage.setItem("Cart", JSON.stringify(Cart));
  localStorage.setItem("address", JSON.stringify(address));
  localStorage.setItem("payMethod", Method);
  localStorage.setItem("ProductsAmount", JSON.stringify(productsAmount));
  const contextValue = {
    state,
    dispatch,
  };
  ExtraAmount = 0;
  return render(
    <>
      <AuthContext.Provider value={contextValue}>
        <Router history={History}>
          <PlaceOrder />
        </Router>
      </AuthContext.Provider>
    </>
  );
};

describe("Check Redirects", () => {
  it("Return to cart Page Removed Values", () => {
    localStorage.clear();
    Setup();
    expect(History.location.pathname).toBe("/");
  });
  it("Return to cart Page on Empty Cart", () => {
    localStorage.setItem("cart", JSON.stringify([]));
    localStorage.setItem("address", JSON.stringify(address));
    localStorage.setItem("payMethod", Method);
    localStorage.setItem("productsAmount", JSON.stringify(productsAmount));
    Setup();
    expect(History.location.pathname).toBe("/");
  });
});

describe("Check Order Details", () => {
  it("Check For Amount Summery", () => {
    Setup();
    const shipping = addDecimals(productsAmount > 500 ? 0 : 100);
    const Tax = addDecimals(productsAmount * 0.1);

    expect(screen.getByTestId("ShippingCost").textContent).toMatch(
      String(shipping)
    );
    expect(screen.getByTestId("TaxCost").textContent).toMatch(String(Tax));

    const TotalAmount = Math.round(shipping + Tax + productsAmount);
    expect(screen.getByTestId("TotalAmount").textContent).toMatch(
      String(TotalAmount)
    );
  });

  it("Store Order in Local Storage And Redirect to Payment Gateway", () => {
    Setup();
    jest.useFakeTimers();
    const shipping = addDecimals(productsAmount > 500 ? 0 : 100);
    const Tax = addDecimals(productsAmount * 0.1);
    const Order = {
      orderItems: Cart,
      shippingAddress: address,
      paymentMethod: Method,
      itemsPrice: productsAmount,
      taxPrice: Tax,
      shippingPrice: shipping,
      totalPrice: Math.round(ExtraAmount + productsAmount),
    };
    const PlaceOrderBtn = screen.getByText(/PlaceOrder/);
    userEvent.click(PlaceOrderBtn);
    act(() => {
      jest.advanceTimersByTime(1);
    });
    expect(PlaceOrderBtn).toBeDisabled();
    expect(JSON.parse(localStorage.order)).toStrictEqual(Order);
    expect(History.location.pathname).toMatch(/PayPal/);
  });
});
