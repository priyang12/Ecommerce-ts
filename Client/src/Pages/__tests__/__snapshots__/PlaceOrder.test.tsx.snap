// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Order Details Check SnapShot of Order 1`] = `
<section
  className="container"
  id="order"
>
  <div
    className="OrderDetails"
  >
    <div
      className="detail"
    >
      <h1>
        SHIPPING 
      </h1>
      <p>
        Address: 
        202,Pipload
         , 
        Surat
         ,
        456123
        , 
        India
      </p>
    </div>
    <div
      className="detail"
    >
      <h1>
        PAYMENT METHOD
      </h1>
      <p>
        Method: 
        PayPal or Credit Card
      </p>
    </div>
    <div
      className="order-details"
    >
      <h1>
        ORDER ITEMS
      </h1>
      <ul
        className="order-list"
      >
        <ul
          className="list"
        >
          <li
            className="list-item"
          >
            <img
              alt="Image"
              src="/Photos/image-1627384388351.webp"
            />
            <a
              href="/product/60d5e622e5179e2bb44bd838"
              onClick={[Function]}
            >
              Airpods Wireless Bluetooth Headphones
            </a>
            <p>
              89.99
            </p>
            <div
              className="quantity"
            >
              <label>
                Qty
              </label>
              <div>
                2
              </div>
            </div>
          </li>
        </ul>
        <ul
          className="list"
        >
          <li
            className="list-item"
          >
            <img
              alt="Image"
              src="/Photos/image-1627385386692.webp"
            />
            <a
              href="/product/60d5e622e5179e2bb44bd83c"
              onClick={[Function]}
            >
              Logtech mouse
            </a>
            <p>
              929.99
            </p>
            <div
              className="quantity"
            >
              <label>
                Qty
              </label>
              <div>
                1
              </div>
            </div>
          </li>
        </ul>
      </ul>
    </div>
  </div>
  <div
    className="OrderSummary"
  >
    <form
      data-testid="PlaceOrder"
      onSubmit={[Function]}
    >
      <ul
        className="checkout"
      >
        <h1>
          ORDER SUMMARY
        </h1>
        <li>
          Items Cost : 
          <span>
             
            2000
          </span>
        </li>
        <li>
          Shipping Cost :
          <span
            data-testid="ShippingCost"
          >
            0
          </span>
        </li>
        <li>
          Tax Cost :
          <span
            data-testid="TaxCost"
          >
            200
          </span>
        </li>
        <li>
          Total Cost :
          <span
            data-testid="TotalAmount"
          >
            2200
          </span>
        </li>
        <li />
      </ul>
      <input
        className="btn"
        disabled={false}
        onClick={[Function]}
        type="submit"
        value="Place Order"
      />
    </form>
  </div>
</section>
`;
